api/agent/state_trackers.go

```diff
@@ -129,7 +131,31 @@ func (c *containerState) GetState() string {
 	return containerStateKeys[res]
 }
 
-func (c *containerState) UpdateState(ctx context.Context, newState ContainerStateType, slots *slotQueue) {
+//This lets the metrics know if a hot container is currently being run for a given app/fn/Image
+func setHot(ctx context.Context, appId string, fnId string, imageName string, state ContainerStateType) {
+	if state != ContainerStateStart && state != ContainerStateDone {
+		//if the container isn't being made or closed then we don't want to change the note of if it's hot or not
+		return
+	}
+
+	ctx, _ = tag.New(ctx,
+		tag.Upsert(appIdKey, appId),
+		tag.Upsert(functionIdKey, fnId),
+		tag.Upsert(imageNameKey, imageName))
```

Comment:
might we do the tag insertion in `UpdateState`? this would apply it for all the container state metrics, which seems useful (i am curious why they seem missing to begin with, but alas). image name seems gratuitous, as fn id points to image name, however in case of function updates it may well be useful to see functions of same id with different images, to see if they hang around, so I think we should leave it (just bringing it up). 

it also seems like this metric is a dupe of a sum of `fn_container_start` + `fn_container_idle` + `fn_container_paused` + `fn_container_busy` - is this equation prohibitive or incorrect for your purposes? I agree that the proposal reads much better out of the box, I'm wondering if we can somehow surface this better using existing metrics, though?

Reply:
Moving the tag insertion to the beginning of UpdateState sounds great.

The problem I ran into with having it work like `fn_container_start` and `fn_container_pause` is that once a metric is added to opencensus I saw no way to remove it, only modify it. So, having start, idle, paused, and busy would require 4 lines, 3 of which would be meaningless at any given point. The other option would be to have `fn_container_status` and have the metric 0-3 for the lifecycle of a container. That isn't prefect either as I believe, and please correct me if I'm wrong, that multiple containers can exist for a given function under high loads. All that said I'm very flexible to changing it if I'm missing something or if you think adding more lines is worth it for clarity purposes.